### DISCUSSION TITLE
`process` is empty object on client side

### DISCUSSION THREAD

**Discussion by PSoltes** [vercel/next.js#43087] [Help]
Hello,

recently i tried disabling polyfills for browser to reduce bundle size of next, eg. `experimental: { fallbackNodePolyfills: false }`. But after enabling it i am not able to access process at client side. On server side eg. in node parts process is available, but in browser process is `{}`. Does anyone have idea why that could be? Adding next.config.js. Console.logging `process` in next.config.js shows correct values.

```
/* eslint-disable import/no-extraneous-dependencies */
// const withBundleAnalyzer = require('@next/bundle-analyzer')
const CaseSensitivePathsPlugin = require('case-sensitive-paths-webpack-plugin')
const nextPluginGraphql = require('nextjs-plugin-graphql')

// const bundleAnalyzer = withBundleAnalyzer({
//   enabled: process.env.ANALYZE === 'true',
// })

const {
  exposeVercelVariables,
} = require('./internal/scripts/vercelSecrets/decrypt')

exposeVercelVariables()

// eslint-disable-next-line no-console,no-process-env
console.log('ENVIRONMENT VARIABLES:', process.env)

const nextConfig = {
  webpack(config) {
    config.plugins.push(new CaseSensitivePathsPlugin())

    return config
  },
  async redirects() {
    return [
      {
        source: '/api-tokens',
        destination: '/api-keys',
        permanent: true,
      },
    ]
  },
  // eslint-disable-next-line no-process-env
  productionBrowserSourceMaps: process.env.ANALYZE === 'true',
  experimental: {
    fallbackNodePolyfills: false,
  },
  eslint: {
    ignoreDuringBuilds: true,
  },
  typescript: {
    ignoreBuildErrors: true,
  },
  images: {
    deviceSizes: [384, 640, 750, 828, 1080, 1200],
  },
}

const plugins = [nextPluginGraphql]

module.exports = () => plugins.reduce((acc, next) => next(acc), nextConfig)
```

**Comment by icyJoseph** [comment]
Weird, process is really a node shim sent to the browser, and in other frameworks it doesn't really exist, for instance vite, https://vitejs.dev/guide/env-and-mode.html

I am surprised because normally Next.js copies the values from the environment and inlines them in the client browser, so process shouldn't be needed. Perhaps some check is missing internally to the webpack configuration. Maybe you could add the process module to webpack as a fallback when the runtime is undefined. 

I'd say ping a maintainer first... it might just be part of the experimental nature of this feature.

**Reply by PSoltes** [reply]
thanks, will experiment more on Monday, really want to have this on, cause bundle reduction is huge

**Reply by icyJoseph** [reply]
Aha, what kind of bundles is this removing? I remember for example that sometimes `crypto` or `Buffer` end up packaged to the final bundle, and to remove those I often have to structure my imports so that client code doesn't end up importing, through many layers of imports, node code, or I just tell webpack to ignore those modules with:

```js
   webpack(config, {nextRuntime}) {
        if (typeof nextRuntime === 'undefined') {
            config.resolve.fallback = {
                ...config.resolve.fallback,
                async_hooks: false, // ignore node async_hooks, which is a node API
            };
            // manually ignore some local modules and node_modules
            const {IgnorePlugin} = require('webpack');
            const ignoreUserModule = new IgnorePlugin({resourceRegExp: /some\/user\/module/});
            const ignoreAWS = new IgnorePlugin({resourceRegExp: /@aws-sdk/});

            config.plugins.push(ignoreUserModule);
            config.plugins.push(ignoreAWS);
        }

        return config;
    },
```

So if I could skip all of this with a flag... I'd be happy. Are you using `@next/bundle-analyzer`?

**Reply by PSoltes** [reply]
yeah, biggest one was `crypto-browserify` will check for more

**Reply by PSoltes** [reply]
@icyJoseph do you know which maintainer should i ping? cause i can't really find a solution how to pass env to client otherwise. Also any idea how to remove this polyfill without this experimental config? tried the IgnorePlugin, but having trouble target this polyfill, as I would have to target crypto and we are using crypto on server bundle.

**Reply by icyJoseph** [reply]
I think regarding maintainers it might be better to open an issue. 

> as I would have to target crypto and we are using crypto on server bundle

That's why you'd only do this if the `nextRuntime` is undefined. `nextRuntime` can be, `undefined | "nodejs" | "edge"` so by doing the typeof check, you tell webpack to not bundle these modules for the client. 

Even more awesome is that you could skip the entire module by changing the `resolve.fallback` object, like I do with `async_hooks` for example.

**Reply by PSoltes** [reply]
ohhh, didn't notice that if. Thanks, but I solved it by enabling improved tree shaking for webpack - `"sideEffects": false` in package.json, as our code is side effect free and can be cut down. Though I will add the fallback thing too, looks nice and will guarantee it not popping in there. I will create an issue too.

**Reply by icyJoseph** [reply]
Yeah, often structuring the imports and modules correctly is enough, like maybe you import a constant string into a React component, from an otherwise heavily focused Node module (such that uses fs, crypto, etc), but that ends up bringing not only one tree, but the entire forest with. 

At the end of the day, for this kind of sensitivity for bundle sizes, it is often best to make sure it can never happen by declaring fallbacks to false for the desired runtime, or straight-up tell Webpack to ignore those paths.
